compiler/rustc_target/src/abi/mod.rs

```diff
@@ -1495,9 +1495,11 @@ impl<'a, Ty> TyAndLayout<'a, Ty> {
     ///
     /// This code is intentionally conservative, and will not detect
     /// * zero init of an enum whose 0 variant does not allow zero initialization
-    /// * making uninitialized types who have a full valid range (ints, floats, raw pointers)
```

Comment:
Ah, I forgot about alignment... it would be handled seamlessly here if we used `0x00..align` as the niche, but right now we only have `0x00..0x01` so we'd not complain about e.g. `&i32`.

IMO the best way to fix that is to make the niche bigger...

Reply:
It's not clear to me why the niche matters, when a 0x1-filled reference means it has pointer 0x0101_0101_0101_0101.

Reply:
We might want to extend the niche of references from the current `0x00..0x01` to `0x00..align` in the future, which would produce invalid values even with `0x01` filling.

Reply:
might_permit_raw_init uses the same information that is used for niches to determine whether a given value is valid for a given type.

Currently, that information says that 0x0101_0101_0101_0101 is valid for `&i32`, which leads to might_permit_raw_init allowing `&i32` to be left uninit with 0x01-filling.

If we extend the niche to take into account alignment, that information will say that 0x0101_0101_0101_0101 is *not* valid for `&i32`, and hence trying to create a 0x01-filled `&i32`  would be rejected (as it should).

Reply:
After spelling this out I realized I am talking nonsense, 0x0101_0101_0101_0101 would still be outside the niche even if `&i32` gets a niche of `0..4`.

So this will need more targeted treatment in `might_permit_raw_init` anyway. Which needs a larger refactor since it neds access to type information. I think either @Nilstrieb or @5225225 already had a version of `might_permit_raw_init`  that takes into account reference type information somewhere.

Reply:
Ok, but those extended niche values are still far from the filled value we're talking about, no?

Reply:
If we had a way to represent a niche that is "all the unaligned values", then that would be sufficient for this check -- 0x0101_0101_0101_0101 would be in the niche and hence disallowed for `&i32`.

But indeed we cannot even represent such niches, so contrary to what I said earlier we cannot rely on niches to detect the trouble around references.

Reply:
Alright, my last reply was having only seen what @Nilstrieb wrote - I guess GitHub failed to dynamically load @RalfJung's replies - but I think we're on the same page now.

So, is this something we should still worry about for LLVM UB in this context? We do tell LLVM about `align`, and their docs say, "If the pointer value does not have the specified alignment, [poison value](https://llvm.org/docs/LangRef.html#poisonvalues) is returned or passed instead." But AIUI that's not UB until it's *used* in a particular way, so is it still a problem here?

Reply:
>  I think either `@Nilstrieb` or @5225225 already had a version of might_permit_raw_init that takes into account reference type information somewhere.

https://github.com/rust-lang/rust/blob/a6ab54bcb5a53bc9479a33de8dddc7fc58f960b1/compiler/rustc_middle/src/ty/layout.rs#L3570-L3584

This? It's in my https://github.com/rust-lang/rust/pull/99389 PR where we allowed 1-aligned references with no `dereferencable` (inside arrays only)

Reply:
@cuviper poison value basically means uninit memory, which the 0x01-filling is intended to avoid, so I am leaning towards saying that this is LLVM UB we should avoid.

@5225225 yes, that one, thanks!